10X 5P scRNA-seq of gd T Cells: Processing
================

``` r
library(Seurat)
library(harmony)
library(SCpubr)
library(scCustomize)
library(tidyverse)
library(future.apply)
library(scRepertoire)
```

# Data Processing

## Reading in Data

``` r
sample_dir <- list.dirs("../../data_raw/10X", recursive=T)
sample_dir <- sample_dir[grepl("VD", sample_dir)]

process_sample <- function(sample_dir) {
  
  batch <- str_extract(sample_dir, "R[1-2]")
  sample_name <- str_extract(sample_dir, "VD[^/]+")
  sample_data <- Read10X(data.dir=sample_dir)
  
  rownames(sample_data$'Antibody Capture') <- sub(
    "^Hu\\.Hashtag_", "hash", rownames(sample_data$'Antibody Capture'))
  
  seurat_obj <- CreateSeuratObject(sample_data$'Gene Expression', project=sample_name)
  
  seurat_obj[["HTO"]] <- CreateAssayObject(sample_data$
    'Antibody Capture'[rownames(sample_data$'Antibody Capture') %in% paste0("hash",1:6), ])
  
  seurat_obj$batch <- batch
  
  seurat_obj <- RenameCells(seurat_obj, add.cell.id=paste0(batch, "_",sample_name))
  
  return(seurat_obj)
}

# workers = number of cores available
plan(multisession, workers=8)
all <- future_lapply(sample_dir, process_sample)
plan(sequential)

rm(sample_dir, process_sample)
```

## Demultiplexing (based on Hashing)

``` r
demux <- function(seurat_obj){
  DefaultAssay(seurat_obj) <- "HTO"
  seurat_obj <- seurat_obj%>%NormalizeData(normalization.method="CLR")%>%
    HTODemux(positive.quantile=.99)
  
  print(summary(as.factor(seurat_obj@meta.data$HTO_classification.global)))
  
  seurat_obj <- subset(seurat_obj, subset=HTO_classification.global=="Singlet")
  
   DefaultAssay(seurat_obj) <- "RNA"
  
  return(seurat_obj)
}

# workers = number of cores available
plan(multisession, workers=8)
all <- future_lapply(all, demux, future.seed=1337)
plan(sequential)

rm(demux)
```

## Adding Meta Data/Removing Unwanted genes

``` r
meta_function <- function(seurat_obj) {
  seurat_obj@meta.data <- seurat_obj@meta.data%>%
    mutate(
      disease=case_when(
        (batch=="R1" & hash.ID %in% c("hash1", "hash2", "hash3")) |
        (batch=="R2" & hash.ID %in% c("hash1", "hash2")) ~ "HD",
        .default="UC"
      )
    )%>%
      unite("diseaseID", disease, batch, hash.ID, sep="_", remove=FALSE)%>%
    mutate(diseaseID=str_replace(diseaseID, pattern="hash", replacement=""))

  seurat_obj$percent.mt <- PercentageFeatureSet(seurat_obj, pattern="^MT-")
  
  # Remove mitochondrial/ribosomal/un-annoted genes from dataset
  genes_to_remove <- grep(rownames(seurat_obj), pattern="^MT-|^RP[SL]|^ENSG[0-9]", value=T)
  seurat_obj <- subset(seurat_obj, features=setdiff(rownames(seurat_obj), genes_to_remove))
  Idents(seurat_obj) <- "all"

  return(seurat_obj)
}

all <- lapply(all, meta_function)

for (i in seq_along(all)){
  Idents(all[[i]]) <- "all"
}
```

## Merging objects (split by gd T subsets)

``` r
vd1 <- merge(all[[1]], all[[4]])%>%JoinLayers()
vd2 <- merge(all[[3]], all[[6]])%>%JoinLayers()

rm(i, meta_function, all)
```

# Vd1

## QC

``` r
VlnPlot_scCustom(vd1, features=c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size=0)
```

<img src="10X_files/figure-gfm/unnamed-chunk-6-1.png" style="display: block; margin: auto;" />

``` r
vd1 <- subset(vd1, subset=nFeature_RNA>500 & nFeature_RNA<3000 &
              nCount_RNA<10000 & percent.mt<10)
```

## Processing & Clustering

``` r
vd1 <- vd1%>%NormalizeData()%>%FindVariableFeatures()%>%ScaleData()%>%RunPCA()%>%
  RunUMAP(dims=1:30)%>%FindNeighbors(dims=1:30)%>%FindClusters()
```

``` r
do_DimPlot(vd1, group.by="batch")
```

<img src="10X_files/figure-gfm/unnamed-chunk-9-1.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd1, group.by="seurat_clusters", label=T)&NoLegend()
```

<img src="10X_files/figure-gfm/unnamed-chunk-9-2.png" style="display: block; margin: auto;" />

``` r
FeaturePlot_scCustom(vd1, features=c("MS4A1", "MS4A7", "CD14", "PPBP", "TRDV1","CD3E"), order=F)&
  NoAxes()&NoLegend()
```

<img src="10X_files/figure-gfm/unnamed-chunk-10-1.png" style="display: block; margin: auto;" />

``` r
VlnPlot_scCustom(vd1, features=c("nFeature_RNA", "percent.mt"), pt.size=0)
```

<img src="10X_files/figure-gfm/unnamed-chunk-11-1.png" style="display: block; margin: auto;" />

``` r
vd1 <- subset(vd1, subset=seurat_clusters %in% c("11", "21", "22", "24", "23", 
                                                 "13", "7", "17"), invert=T)
```

## Batch Correction (Batch + DonorID)

``` r
vd1 <- vd1%>%FindVariableFeatures()%>%ScaleData()%>%RunPCA()%>%
  RunHarmony(c("batch", "diseaseID"))
```

``` r
ElbowPlot(vd1, reduction="harmony", 50)
```

<img src="10X_files/figure-gfm/unnamed-chunk-14-1.png" style="display: block; margin: auto;" />

``` r
vd1 <- vd1%>%RunUMAP(reduction="harmony", dims=1:20)%>%
  FindNeighbors(reduction="harmony", dims=1:20)%>%FindClusters(resolution=.4)
```

``` r
do_DimPlot(vd1, group.by="batch")
```

<img src="10X_files/figure-gfm/unnamed-chunk-16-1.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd1, group.by="disease")
```

<img src="10X_files/figure-gfm/unnamed-chunk-16-2.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd1, group.by="diseaseID")
```

<img src="10X_files/figure-gfm/unnamed-chunk-16-3.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd1, group.by="seurat_clusters", label=T)&NoLegend()
```

<img src="10X_files/figure-gfm/unnamed-chunk-16-4.png" style="display: block; margin: auto;" />
\## TCR

``` r
contig <- list("R1"=read_tsv( "../../data_raw/10X/R1/VD1/clones.tsv"),
               "R2"=read_tsv("../../data_raw/10X/R2/VD1/clones.tsv"))
contig <- loadContigs(contig, format = "MiXCR")

# Remove allele variant calls + scoring (*00(###))
contig <- lapply(contig, function(df) {
    df %>%
    mutate_at(paste0(c("v", "d", "j", "c"), "_gene"), ~str_replace(., "\\*.*", ""))
})

combined <- combineTCR(contig,  removeNA=T, filterMulti=T)
names(combined) <- names(contig)

for (i in seq_along(combined)) {
  combined[[i]]$barcode <- paste0(names(combined[i]), "_VD1_", combined[[i]]$barcode, "-1")
}

combined <- list(bind_rows(combined))

vd1 <- combineExpression(combined, vd1, proportion=F,cloneSize=c(
  Single=1, Small=5, Medium=25, Large=100, Hyper=100
))

rm(i, combined, contig)
```

# Vd2

## QC

``` r
VlnPlot_scCustom(vd2, features=c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size=0)
```

<img src="10X_files/figure-gfm/unnamed-chunk-18-1.png" style="display: block; margin: auto;" />

``` r
vd2 <- subset(vd2, subset=nFeature_RNA>200 & nFeature_RNA<3000 & percent.mt<15)
```

## Processing & Clustering

``` r
vd2 <- vd2%>%NormalizeData()%>%FindVariableFeatures()%>%ScaleData()%>%RunPCA()%>%
  RunUMAP(dims=1:30)%>%FindNeighbors(dims=1:30)%>%FindClusters()
```

``` r
do_DimPlot(vd2, group.by="seurat_clusters", label=T)&NoLegend()
```

<img src="10X_files/figure-gfm/unnamed-chunk-21-1.png" style="display: block; margin: auto;" />

``` r
FeaturePlot_scCustom(vd2, features=c("MS4A1", "MS4A7", "CD14", "PPBP", "CD3E", "TRDV2"), order=F)&
  NoLegend()&NoAxes()
```

<img src="10X_files/figure-gfm/unnamed-chunk-21-2.png" style="display: block; margin: auto;" />

``` r
VlnPlot_scCustom(vd2, features=c("nFeature_RNA", "percent.mt", "CD3E"), pt.size=0)&
  geom_boxplot(fill="white", width=.5, coef=0, outlier.shape=NA)
```

<img src="10X_files/figure-gfm/unnamed-chunk-22-1.png" style="display: block; margin: auto;" />

``` r
# Remove Contaminating Cells/low quality Cells
vd2 <- subset(vd2, subset=seurat_clusters %in% c("4", "8", "14", "16"), invert=T)
```

## Batch Correction (Batch + DonorID)

``` r
vd2 <- vd2%>%FindVariableFeatures()%>%ScaleData()%>%RunPCA()%>%
  RunHarmony(c("batch", "diseaseID"))
```

``` r
ElbowPlot(vd2, reduction="harmony", 50)
```

<img src="10X_files/figure-gfm/unnamed-chunk-25-1.png" style="display: block; margin: auto;" />

``` r
vd2 <- vd2%>%RunUMAP(reduction="harmony", dims=1:20)%>%
  FindNeighbors(reduction="harmony", dims=1:20)%>%FindClusters()
```

``` r
do_DimPlot(vd2, group.by="batch")
```

<img src="10X_files/figure-gfm/unnamed-chunk-27-1.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd2, group.by="disease")
```

<img src="10X_files/figure-gfm/unnamed-chunk-27-2.png" style="display: block; margin: auto;" />

``` r
do_DimPlot(vd2, group.by="seurat_clusters", label=T)&NoLegend()
```

<img src="10X_files/figure-gfm/unnamed-chunk-27-3.png" style="display: block; margin: auto;" />

## TCR

``` r
contig <- list("R1"=read_tsv( "../../data_raw/10X/R1/VD2/clones.tsv"),
               "R2"=read_tsv("../../data_raw/10X/R2/VD2/clones.tsv"))
contig <- loadContigs(contig, format = "MiXCR")

# Remove allele variant calls + scoring (*00(###))
contig <- lapply(contig, function(df) {
    df %>%
    mutate_at(paste0(c("v", "d", "j", "c"), "_gene"), ~str_replace(., "\\*.*", ""))
})

combined <- combineTCR(contig,  removeNA=T, filterMulti=T)
names(combined) <- names(contig)

for (i in seq_along(combined)) {
  combined[[i]]$barcode <- paste0(names(combined[i]), "_VD2_", combined[[i]]$barcode, "-1")
}

combined <- list(bind_rows(combined))

vd2 <- combineExpression(combined, vd2, proportion=F,cloneSize=c(
  Single=1, Small=5, Medium=25, Large=100, Hyper=100
))

rm(i, combined, contig)
```

# Saving Objects

``` r
saveRDS(vd1, "../../data_processed/10X/Vd1.Rds")
saveRDS(vd2, "../../data_processed/10X/Vd2.Rds")
```

# Session Info

``` r
sessionInfo()
```

    ## R version 4.4.0 (2024-04-24)
    ## Platform: x86_64-pc-linux-gnu
    ## Running under: Ubuntu 22.04.4 LTS
    ## 
    ## Matrix products: default
    ## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
    ## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0
    ## 
    ## locale:
    ##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
    ##  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    
    ##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8   
    ##  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
    ##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
    ## [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
    ## 
    ## time zone: Europe/Berlin
    ## tzcode source: system (glibc)
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ##  [1] scRepertoire_2.0.8  future.apply_1.11.2 future_1.33.2      
    ##  [4] lubridate_1.9.3     forcats_1.0.0       stringr_1.5.1      
    ##  [7] dplyr_1.1.4         purrr_1.0.2         readr_2.1.5        
    ## [10] tidyr_1.3.1         tibble_3.2.1        ggplot2_3.5.1      
    ## [13] tidyverse_2.0.0     scCustomize_2.1.2   SCpubr_2.0.2       
    ## [16] harmony_1.2.0       Rcpp_1.0.12         Seurat_5.1.0       
    ## [19] SeuratObject_5.0.2  sp_2.1-4           
    ## 
    ## loaded via a namespace (and not attached):
    ##   [1] fs_1.6.4                    matrixStats_1.3.0          
    ##   [3] spatstat.sparse_3.0-3       httr_1.4.7                 
    ##   [5] RColorBrewer_1.1-3          tools_4.4.0                
    ##   [7] sctransform_0.4.1           utf8_1.2.4                 
    ##   [9] R6_2.5.1                    lazyeval_0.2.2             
    ##  [11] uwot_0.2.2                  withr_3.0.0                
    ##  [13] gridExtra_2.3               progressr_0.14.0           
    ##  [15] quantreg_5.97               cli_3.6.2                  
    ##  [17] Biobase_2.64.0              spatstat.explore_3.2-7     
    ##  [19] fastDummies_1.7.3           iNEXT_3.0.1                
    ##  [21] labeling_0.4.3              prismatic_1.1.2            
    ##  [23] spatstat.data_3.0-4         ggridges_0.5.6             
    ##  [25] pbapply_1.7-2               yulab.utils_0.1.4          
    ##  [27] stringdist_0.9.12           parallelly_1.37.1          
    ##  [29] VGAM_1.1-11                 rstudioapi_0.16.0          
    ##  [31] generics_0.1.3              gridGraphics_0.5-1         
    ##  [33] shape_1.4.6.1               ica_1.0-3                  
    ##  [35] spatstat.random_3.2-3       vroom_1.6.5                
    ##  [37] Matrix_1.6-5                ggbeeswarm_0.7.2           
    ##  [39] fansi_1.0.6                 S4Vectors_0.42.0           
    ##  [41] abind_1.4-5                 lifecycle_1.0.4            
    ##  [43] yaml_2.3.8                  snakecase_0.11.1           
    ##  [45] SummarizedExperiment_1.34.0 SparseArray_1.4.5          
    ##  [47] Rtsne_0.17                  paletteer_1.6.0            
    ##  [49] grid_4.4.0                  promises_1.3.0             
    ##  [51] crayon_1.5.2                miniUI_0.1.1.1             
    ##  [53] lattice_0.22-5              cowplot_1.1.3              
    ##  [55] pillar_1.9.0                knitr_1.46                 
    ##  [57] GenomicRanges_1.56.0        rjson_0.2.21               
    ##  [59] codetools_0.2-19            leiden_0.4.3.1             
    ##  [61] glue_1.7.0                  data.table_1.15.4          
    ##  [63] vctrs_0.6.5                 png_0.1-8                  
    ##  [65] spam_2.10-0                 gtable_0.3.5               
    ##  [67] assertthat_0.2.1            rematch2_2.1.2             
    ##  [69] cachem_1.1.0                xfun_0.44                  
    ##  [71] S4Arrays_1.4.1              mime_0.12                  
    ##  [73] tidygraph_1.3.1             survival_3.7-0             
    ##  [75] SingleCellExperiment_1.26.0 fitdistrplus_1.1-11        
    ##  [77] ROCR_1.0-11                 nlme_3.1-165               
    ##  [79] bit64_4.0.5                 RcppAnnoy_0.0.22           
    ##  [81] evd_2.3-7                   GenomeInfoDb_1.40.0        
    ##  [83] irlba_2.3.5.1               vipor_0.4.7                
    ##  [85] KernSmooth_2.23-24          colorspace_2.1-0           
    ##  [87] BiocGenerics_0.50.0         ggrastr_1.0.2              
    ##  [89] tidyselect_1.2.1            bit_4.0.5                  
    ##  [91] compiler_4.4.0              SparseM_1.81               
    ##  [93] ggdendro_0.2.0              DelayedArray_0.30.1        
    ##  [95] plotly_4.10.4               scales_1.3.0               
    ##  [97] lmtest_0.9-40               digest_0.6.35              
    ##  [99] goftest_1.2-3               spatstat.utils_3.0-4       
    ## [101] rmarkdown_2.27              XVector_0.44.0             
    ## [103] RhpcBLASctl_0.23-42         htmltools_0.5.8.1          
    ## [105] pkgconfig_2.0.3             MatrixGenerics_1.16.0      
    ## [107] highr_0.10                  fastmap_1.2.0              
    ## [109] rlang_1.1.3                 GlobalOptions_0.1.2        
    ## [111] htmlwidgets_1.6.4           UCSC.utils_1.0.0           
    ## [113] shiny_1.8.1.1               farver_2.1.2               
    ## [115] zoo_1.8-12                  jsonlite_1.8.8             
    ## [117] magrittr_2.0.3              GenomeInfoDbData_1.2.12    
    ## [119] ggplotify_0.1.2             dotCall64_1.1-1            
    ## [121] patchwork_1.2.0             munsell_0.5.1              
    ## [123] evmix_2.12                  viridis_0.6.5              
    ## [125] reticulate_1.37.0           truncdist_1.0-2            
    ## [127] stringi_1.8.4               ggalluvial_0.12.5          
    ## [129] ggraph_2.2.1                zlibbioc_1.50.0            
    ## [131] MASS_7.3-60                 plyr_1.8.9                 
    ## [133] parallel_4.4.0              listenv_0.9.1              
    ## [135] ggrepel_0.9.5               deldir_2.0-4               
    ## [137] graphlayouts_1.1.1          splines_4.4.0              
    ## [139] tensor_1.5                  hash_2.2.6.3               
    ## [141] hms_1.1.3                   circlize_0.4.16            
    ## [143] igraph_2.0.3                spatstat.geom_3.2-9        
    ## [145] cubature_2.1.0              RcppHNSW_0.6.0             
    ## [147] reshape2_1.4.4              stats4_4.4.0               
    ## [149] evaluate_0.23               ggprism_1.0.5              
    ## [151] tzdb_0.4.0                  tweenr_2.0.3               
    ## [153] httpuv_1.6.15               MatrixModels_0.5-3         
    ## [155] RANN_2.6.1                  polyclip_1.10-6            
    ## [157] scattermore_1.2             ggforce_0.4.2              
    ## [159] janitor_2.2.0               xtable_1.8-4               
    ## [161] RSpectra_0.16-1             later_1.3.2                
    ## [163] viridisLite_0.4.2           gsl_2.1-8                  
    ## [165] memoise_2.0.1               beeswarm_0.4.0             
    ## [167] IRanges_2.38.0              cluster_2.1.6              
    ## [169] timechange_0.3.0            globals_0.16.3
